In [8]:
import csv
import xlrd
import pandas as pd

# https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )


In [4]:
annual_geographical_mobility_rates_type_movement

Table with row headers in column A and column headers in rows 5 through 7.  \
0    Table A-1. Annual Geographical Mobility Rates,...                           
1                              (Numbers in thousands.)                           
2                                                  NaN                           
3                                      Mobility period                           
4                                                  NaN                           
5                                                  NaN                           
6                                               NUMBER                           
7                                                  NaN                           
8                                            2017-2018                           
9                                            2016-2017                           
10                                           2015-2016                           
11                                                 NaN                           
12                                           2014-2015                           
13                   2013-2014 (98,000 address file)/1                           
14                                           2012-2013                           
15                                           2011-2012                           
16                         2010-2011 (2010 controls)/2                           
17                         2010-2011 (2000 controls)/3                           
18                                                 NaN                           
19                         2009-2010 (2010 controls)/2                           
20                         2009-2010 (2000 controls)/3                           
21                                           2008-2009                           
22                                           2007-2008                           
23                                           2006-2007                           
24                                           2005-2006                           
25                                                 NaN                           
26                                         2004-2005/4                           
27                                           2003-2004                           
28                                           2002-2003                           
29                                           2001-2002                           
..                                                 ...                           
156                                          1960-1961                           
157                                                NaN                           
158                                          1959-1960                           
159                                          1958-1959                           
160                                          1957-1958                           
161                                          1956-1957                           
162                                          1955-1956                           
163                                                NaN                           
164                                          1954-1955                           
165                                          1953-1954                           
166                                          1952-1953                           
167                                          1951-1952                           
168                                          1950-1951                           
169                                                NaN                           
170                                          1949-1950                           
171                                          1948-1949                           
172                                          1947-1948              

annual_geographical_mobility_rates_type_movement

notes:
    seems to cover back to 1947
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [5]:
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions

Table with row headers in column A and column headers in row 5.  \
0    Table A-2.  Annual Inmigration, Outmigration, ...                
1                              (Numbers in thousands.)                
2                     (leading dot indicates sub-part)                
3                Mobility period and type of migration                
4                                                  NaN                
5                                            2017-2018                
6                                          .Inmigrants                
7                                         .Outmigrants                
8                              .Net internal migration                
9                                  .Movers from abroad                
10                   .Net migration (including abroad)                
11                                                 NaN                
12                                           2016-2017                
13                                         .Inmigrants                
14                                        .Outmigrants                
15                             .Net internal migration                
16                                 .Movers from abroad                
17                   .Net migration (including abroad)                
18                                                 NaN                
19                                           2015-2016                
20                                         .Inmigrants                
21                                        .Outmigrants                
22                             .Net internal migration                
23                                 .Movers from abroad                
24                   .Net migration (including abroad)                
25                                                 NaN                
26                                           2014-2015                
27                                         .Inmigrants                
28                                        .Outmigrants                
29                             .Net internal migration                
..                                                 ...                
272                                                NaN                
273                                          1982-1983                
274                                        .Inmigrants                
275                                       .Outmigrants                
276                            .Net internal migration                
277                                .Movers from abroad                
278                  .Net migration (including abroad)                
279                                                NaN                
280                                          1981-1982                
281                                        .Inmigrants                
282                                       .Outmigrants                
283                            .Net internal migration                
284                                .Movers from abroad                
285                  .Net migration (including abroad)                
286                                                NaN                
287                                          1980-1981                
288                                        .Inmigrants                
289                                       .Outmigrants                
290                            .Net internal migration                
291                                .Movers from abroad                
292                  .Net migration (including abroad)                
293                                         Footnotes:                
294  * Net flow significantly different from zero a...                
295  1/ The 2014 CPS ASEC included redesigned quest...                
296  2/ Population controls 

annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions

notes:
    seems to cover back to 1980
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [6]:
inmigration_outmigration__net_migration_metropolitan_status

Table with row headers in column A and column headers in row 5.  \
0    Table A-3.  Inmigration, Outmigration, and Net...                
1                              (Numbers in thousands.)                
2                    (Leading dots indicate sub-parts)                
3              Mobility period and Metropolitan status                
4                    (leading dots indicate sub-parts)                
5                                            2017-2018                
6                                  .Metropolitan areas                
7                                   ..Principal cities                
8                   ..Balances of metropolitan areas/8                
9                               .Nonmetropolitan areas                
10                                                 NaN                
11                                           2016-2017                
12                                 .Metropolitan areas                
13                                  ..Principal cities                
14                                           ..Suburbs                
15                              .Nonmetropolitan areas                
16                                                 NaN                
17                                           2015-2016                
18                                 .Metropolitan areas                
19                                  ..Principal cities                
20                                           ..Suburbs                
21                              .Nonmetropolitan areas                
22                                                 NaN                
23                                         2014-2015/1                
24                                                 NaN                
25                   2013-2014 (98,000 address file)/2                
26                                 .Metropolitan areas                
27                                  ..Principal cities                
28                                           ..Suburbs                
29                              .Nonmetropolitan areas                
..                                                 ...                
182                                .Metropolitan areas                
183                                   ..Central cities                
184                                          ..Suburbs                
185                             .Nonmetropolitan areas                
186                                                NaN                
187                                          1986-1987                
188                                .Metropolitan areas                
189                                   ..Central cities                
190                                          ..Suburbs                
191                             .Nonmetropolitan areas                
192                                                NaN                
193                                          1985-1986                
194                                .Metropolitan areas                
195                                   ..Central cities                
196                                          ..Suburbs                
197                             .Nonmetropolitan areas                
198                                         Footnotes:                
199  * Net flow significantly different from zero a...                
200  1/ Metropolitan mobility data are not availabl...                
201  2/ The 2014 CPS ASEC included redesigned quest...                
202  3/ Population controls consistent with 2010 Ce...                
203  4/ Population controls consistent with 2000 Ce...                
204  5/ Metropolitan mobility data are not availabl...                
205  6/ SCHIP sample: population controls consisten...                
206  7/ 1-year metropolitan 

inmigration_outmigration__net_migration_metropolitan_status

notes:
    seems to cover back to 1985
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [9]:
geographical_mobility_by_tenure

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
0    Table A-4. Geographical Mobility by Tenure:  1...                                                                                                                                 
1                              (Numbers in thousands.)                                                                                                                                 
2                     (leading dot indicates sub-part)                                                                                                                                 
3                           Mobility period and Tenure                                                                                                                                 
4                                                  NaN                                                                                                                                 
5                                                  NaN                                                                                                                                 
6                                              NUMBERS                                                                                                                                 
7                                                  NaN                                                                                                                                 
8                                            2017-2018                                                                                                                                 
9                              .Total, 1 year and over                                                                                                                                 
10                           ..In owner-occupied units                                                                                                                                 
11                          ..In renter-occupied units                                                                                                                                 
12                                                 NaN                                                                                                                                 
13                                           2016-2017                                                                                                                                 
14                             .Total, 1 year and over                                                                                                                                 
15                           ..In owner-occupied units                                                                                                                                 
16                          ..In renter-occupied units                                                                                                                                 
17                                                 NaN                                                                                                                                 
18                                           2015-2016                                                                                                                                 
19                             .Total, 1 year and over                                                                                                                                 
20                           ..In owner-occupied units                                                                                      

geographical_mobility_by_tenure

notes:
    seems to cover back to 1987
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [10]:
reason_for_move_specific_categories

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
0   Table A-5. Reason for Move (Specific Categorie...                                                                                    
1                             (Numbers in thousands.)                                                                                    
2                   (leading dots indicate sub-parts)                                                                                    
3                                     Mobility Period                                                                                    
4                                                 NaN                                                                                    
5                                              NUMBER                                                                                    
6                                                 NaN                                                                                    
7                                         2017-2018/1                                                                                    
8                                         2016-2017/1                                                                                    
9                                         2015-2016/1                                                                                    
10                                          2014-2015                                                                                    
11                  2013-2014 (98,000 address file)/2                                                                                    
12                                          2012-2013                                                                                    
13                                          2011-2012                                                                                    
14                        2010-2011 (2010 controls)/3                                                                                    
15                        2010-2011 (2000 controls)/4                                                                                    
16                                                NaN                                                                                    
17                        2009-2010 (2010 controls)/3                                                                                    
18                        2009-2010 (2000 controls)/4                                                                                    
19                                          2008-2009                                                                                    
20                                          2007-2008                                                                                    
21                                          2006-2007                                                                                    
22                                          2005-2006                                                                                    
23                                                NaN                                                                                    
24                                          2004-2005                                                                                    
25                                          2003-2004                                                                                    
26                                          2002-2003                                                                                    
27                                          2001-2002                                                                                    
2

reason_for_move_specific_categories

notes:
    seems to cover back to 1998
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [11]:
distance_of_intercounty_move

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
0   Table A-6. Distance of Intercounty Move: 2003-...                                                                     
1                             (Numbers in thousands.)                                                                     
2                                                 NaN                                                                     
3                                     Mobility Period                                                                     
4                                              NUMBER                                                                     
5                                                 NaN                                                                     
6                                           2017-2018                                                                     
7                                           2016-2017                                                                     
8                                           2015-2016                                                                     
9                                                 NaN                                                                     
10                                          2014-2015                                                                     
11                  2013-2014 (98,000 address file)/1                                                                     
12                                          2012-2013                                                                     
13                                          2011-2012                                                                     
14                        2010-2011 (2010 controls)/2                                                                     
15                        2010-2011 (2000 controls)/3                                                                     
16                                                NaN                                                                     
17                        2009-2010 (2010 controls)/2                                                                     
18                        2009-2010 (2000 controls)/3                                                                     
19                                          2008-2009                                                                     
20                                          2007-2008                                                                     
21                                          2006-2007                                                                     
22                                          2005-2006                                                                     
23                                                NaN                                                                     
24                                        2004-2005/4                                                                     
25                                        2003-2004/4                                                                     
26                                        2002-2003/4                                                                     
27                                                NaN                                                                     
28                                            PERCENT                                                                     
29                                                NaN                                                                     
30                                          2017-2018                                                                     
31                                          2016-2017              

distance_of_intercounty_move

notes:
    seems to cover back to 2002
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

## Aggregate Notes
inconsistent dates
    - i.e. first recording year varies by up to 55 years 

| first recording year | dataset |
|------|------|
| 1947 | annual_geographical_mobility_rates_type_movement |
|------|------|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|------|------|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1998 | reason_for_move_specific_categories |
|------|------|
| 2002 | distance_of_intercounty_move |



In [31]:
"""
let's drop the rows clustered with NaNs
"""

# rename dataframe
df = annual_geographical_mobility_rates_type_movement
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
df

Table with row headers in column A and column headers in rows 5 through 7.  \
0    Table A-1. Annual Geographical Mobility Rates,...                           
1                              (Numbers in thousands.)                           
3                                      Mobility period                           
4                                                  NaN                           
5                                                  NaN                           
6                                               NUMBER                           
8                                            2017-2018                           
9                                            2016-2017                           
10                                           2015-2016                           
12                                           2014-2015                           
13                   2013-2014 (98,000 address file)/1                           
14                                           2012-2013                           
15                                           2011-2012                           
16                         2010-2011 (2010 controls)/2                           
17                         2010-2011 (2000 controls)/3                           
19                         2009-2010 (2010 controls)/2                           
20                         2009-2010 (2000 controls)/3                           
21                                           2008-2009                           
22                                           2007-2008                           
23                                           2006-2007                           
24                                           2005-2006                           
26                                         2004-2005/4                           
27                                           2003-2004                           
28                                           2002-2003                           
29                                           2001-2002                           
30                         2000-2001 (2000 controls)/3                           
31                  2000-2001 (SCHIP, 2000 controls)/5                           
32                         2000-2001 (1990 controls)/6                           
34                                           1999-2000                           
35                                           1998-1999                           
..                                                 ...                           
153                                          1963-1964                           
154                                          1962-1963                           
155                                          1961-1962                           
156                                          1960-1961                           
158                                          1959-1960                           
159                                          1958-1959                           
160                                          1957-1958                           
161                                          1956-1957                           
162                                          1955-1956                           
164                                          1954-1955                           
165                                          1953-1954                           
166                                          1952-1953                           
167                                          1951-1952                           
168                                          1950-1951                           
170                                          1949-1950                           
171                                          1948-1949                           
172                                          1947-1948              

notes:
    some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    rows 173 - 185 contain only 1 NaN value

actions:
    drop all rows without 2 or more non-NaN values
    merge rows 3 and 